main.go

```diff
@@ -184,56 +190,56 @@ type HTTPError struct {
 }
 
 func (api *API) remoteRead(w http.ResponseWriter, r *http.Request, logger log.Logger) error {
-	ctx := r.Context()
-	tracer := otel.Tracer("")
-	var span trace.Span
-	ctx, span = tracer.Start(ctx, "remoteRead")
-	defer span.End()
-
-	compressed, err := ioutil.ReadAll(r.Body)
-	if err != nil {
-		return err
-	}
-
-	reqBuf, err := snappy.Decode(nil, compressed)
-	if err != nil {
-		return HTTPError{err, http.StatusBadRequest}
-	}
-
-	var req prompb.ReadRequest
-	if err := proto.Unmarshal(reqBuf, &req); err != nil {
-		return HTTPError{err, http.StatusBadRequest}
-	}
-
-	// Ignored selectors.
-	ignoredSelector := make(map[string]struct{})
-	if ignores, ok := r.URL.Query()["ignore"]; ok {
-		for _, ignore := range ignores {
-			ignoredSelector[ignore] = struct{}{}
-		}
-	}
-
-	// This does not do streaming, at the time of writing Prometheus doesn't ask
-	// for it anyway: https://github.com/prometheus/prometheus/issues/5926
-
-	resp, err := api.doStoreRequest(r.Context(), &req, ignoredSelector, logger)
-	if err != nil {
-		return err
-	}
-
-	data, err := proto.Marshal(resp)
-	if err != nil {
-		return err
-	}
-
-	w.Header().Set("Content-Type", "application/x-protobuf")
-	w.Header().Set("Content-Encoding", "snappy")
-
-	compressed = snappy.Encode(nil, data)
-	if _, err := w.Write(compressed); err != nil {
-		level.Error(logger).Log("err", err, "traceID", span.SpanContext().TraceID)
-	}
-	return nil
+    ctx := r.Context()
+    tracer := otel.Tracer("")
+    var span trace.Span
+    ctx, span = tracer.Start(ctx, "remoteRead")
+    defer span.End()
+
+    compressed, err := ioutil.ReadAll(r.Body)
+    if err != nil {
+        return err
+    }
+
+    reqBuf, err := snappy.Decode(nil, compressed)
+    if err != nil {
+        return HTTPError{err, http.StatusBadRequest}
+    }
+
+    var req prompb.ReadRequest
+    if err := proto.Unmarshal(reqBuf, &req); err != nil {
+        return HTTPError{err, http.StatusBadRequest}
+    }
+
+    ignoredSelector := make(map[string]struct{})
+    if ignores, ok := r.URL.Query()["ignore"]; ok {
+        for _, ignore := range ignores {
+            ignoredSelector[ignore] = struct{}{}
+        }
+    }
+
+    // Use the updated context `ctx` that includes the span for tracing.
+    resp, err := api.doStoreRequest(ctx, &req, ignoredSelector, logger)
+    if err != nil {
+        return err
+    }
+
+    data, err := proto.Marshal(resp)
+    if err != nil {
+        return err
+    }
+
+    w.Header().Set("Content-Type", "application/x-protobuf")
+    w.Header().Set("Content-Encoding", "snappy")
+
+    compressed = snappy.Encode(nil, data)
+    if _, err := w.Write(compressed); err != nil {
+        if logErr := level.Error(logger).Log("err", err, "traceID", span.SpanContext().TraceID); logErr != nil {
+            fmt.Fprintf(os.Stderr, "original error: %v, logging error: %v\n", err, logErr)
+        }
+        os.Exit(1)
```

Comment:
Was this `os.Exit` intentional? I think this makes the binary exit every time Prometheus aborts the remote read request. E.g.

```
ts=2024-06-26T14:38:07.680190285Z caller=main.go:239 level=error err="write tcp 127.0.0.1:10080->127.0.0.1:51420: write: broken pipe" traceID=00000000000000000000000000000000
```